core/src/main/java/org/apache/calcite/sql/validate/implicit/TypeCoercionImpl.java

```diff
@@ -121,8 +124,8 @@ public TypeCoercionImpl(RelDataTypeFactory typeFactory, SqlValidator validator)
       // Set operations are binary for now.
       final SqlCall operand0 = ((SqlCall) query).operand(0);
       final SqlCall operand1 = ((SqlCall) query).operand(1);
-      final boolean coerced = rowTypeCoercion(scope, operand0, columnIndex, targetType)
-          && rowTypeCoercion(scope, operand1, columnIndex, targetType);
+      boolean coerced = rowTypeCoercion(scope, operand0, columnIndex, targetType);
```

Comment:
Change `&&` directly to `||` ?

Reply:
This will not fix the problem. Because when `rowTypeCoercion(scope, operand0, columnIndex, targetType)=false`,  `rowTypeCoercion(scope, operand1, columnIndex, targetType) `will not be run.
I will use `|`, do you agree?
```
final boolean coerced = rowTypeCoercion(scope, operand0, columnIndex, targetType)
    | rowTypeCoercion(scope, operand1, columnIndex, targetType);
```

Reply:
Yeah, you code is correct, we should not add any short-cut logic here.

Reply:
done